### PR TITLE
Add youtube handler inside google provider

### PR DESCRIPTION
**Pull request by chime3**
### What's changed?
- Add youtube following(Subscription) handler.
- Add youtube post(Activity) handler.

### How to test these changes
Worked on these tests, but pre-built tests are broken now and need to discuss before make changes.

### Anything to be aware of?
- Invalid credential errors. Need to confirm if @tahpot changed credentials.
- `SyncSchemaPosition` type is removed recently and need to confirm if it is removed accidently.

### CODE REVIEW

File: src/providers/google/youtube-following.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            const insertedAt = snippet.publishedAt || "Unknown";
+
+            if (breakTimestamp && insertedAt < breakTimestamp) {
+                break;
+            }
+
+            const title = snippet.title || "No title";
+
+            results.push({
+                _id: `youtube-following-${itemId}`,
```
Comment (tahpot) [inline src/providers/google/youtube-following.ts:160]: `youtube-${itemId}`

File: src/providers/google/youtube-following.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                name: title,
```
Comment (tahpot) [inline src/providers/google/youtube-following.ts:161]: Other fields that should be added if available; icon, uri.

Also populate followedTimestamp and sourceData (see facebook following)

File: src/providers/google/youtube-post.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+            if (breakTimestamp && insertedAt < breakTimestamp) {
+                break;
+            }
+
+            const title = snippet.title || "No title";
+
+            results.push({
+                _id: `youtube-post-${itemId}`,
+                name: title,
```
Comment (tahpot) [inline src/providers/google/youtube-post.ts:161]: See comments on youtube following

Comment (tahpot) [reply src/providers/google/youtube-post.ts:161]: Can you provide an example of all the data available in `item`? I'd like to see if there's more data we can include.

File: src/providers/google/youtube-following.ts (hunk 2 above)
Comment (tahpot) [reply src/providers/google/youtube-following.ts:161]: Can you provide an example of all the data available in `item`? I'd like to see if there's more data we can include.

**Review by tahpot — COMMENTED** [review]
I've added some initial comments. 

Please make sure you have read the [Contributing.md](https://github.com/verida/data-connector-server/blob/develop/docs/Contributing.md) notes and followed them.

File: src/providers/google/youtube-post.ts (hunk 3 above)
Comment (chime3) [reply src/providers/google/youtube-post.ts:193]: > Can you provide an example of all the data available in `item`? I'd like to see if there's more data we can include.

### Youtube activity has multiple types and typical types are:
- upload: Indicates that a user has uploaded a video.
- like: Indicates that a user has liked a video.
- comment: Indicates that a user has commented on a video.
- subscription: Indicates that a user has subscribed to a channel.
- recommendation: Indicates that a user has recommended a video.
- playlistItem: Indicates that a user has added a video to a playlist.

Each type has different `contentDeatils` like this:
```
{
  "snippet": {
    "type": "upload"|"like"|"xxx",
    "channelId": "",
    "title": ",
    "description": "",
    "thumbnails": {
      "default": {
        "url": "https://example.com/default.jpg",
        "width": 120,
        "height": 90
      }
    }
  },
  "contentDetails": {
    "upload": {
      "videoId": "video_id"
    }
    "like": {
      "resourceId": {
        "kind": "youtube#video",
        "videoId": "liked_video_id"
      }
    }
  "xxx": {}
  }
}
```
### Youtube subscription response is static 
```
{
  "kind": "youtube#subscriptionListResponse",
  "etag": "etag",
  "items": [
    {
      "kind": "youtube#subscription",
      "etag": "etag",
      "id": "subscriptionId",
      "snippet": {
        "publishedAt": "2023-07-14T12:34:56.000Z",
        "title": "Channel Title",
        "description": "Channel Description",
        "resourceId": {
          "kind": "youtube#channel",
          "channelId": "channelId"
        },
        "channelId": "subscribedChannelId",
        "thumbnails": {
          "default": {
            "url": "https://example.com/default.jpg",
            "width": 88,
            "height": 88
          },
          "medium": {
            "url": "https://example.com/medium.jpg",
            "width": 240,
            "height": 240
          },
          "high": {
            "url": "https://example.com/high.jpg",
            "width": 800,
            "height": 800
          }
        }
      }
    }
  ]
}
```

File: src/providers/google/index.ts [hunk 4]
```diff
@@ -25,15 +25,26 @@ export default class GoogleProvider extends Base {
...
-      "email",
-      "https://www.googleapis.com/auth/gmail.readonly",
+        "profile",
+        "openid",
+        "email",
+        "https://www.googleapis.com/auth/gmail.readonly",
+        "https://www.googleapis.com/auth/youtube",
+        "https://www.googleapis.com/auth/youtube.channel-memberships.creator",
+        "https://www.googleapis.com/auth/youtube.force-ssl",
+        "https://www.googleapis.com/auth/youtube.readonly",
```
[the remark this answers is not in the document]
Comment (tahpot) [reply src/providers/google/index.ts:43]: Please explain these scopes. Why do we need more than just `youtube.readonly`?

File: src/providers/google/youtube-following.ts (hunk 1 above)
Comment (chime3) [reply src/providers/google/youtube-following.ts]: > `youtube-${itemId}`

Solved

File: src/providers/google/youtube-following.ts (hunk 2 above)
Comment (chime3) [reply src/providers/google/youtube-following.ts:164]: Please check answer below

File: src/providers/google/youtube-post.ts (hunk 3 above)
Comment (tahpot) [reply src/providers/google/youtube-post.ts:193]: Thanks, we will need to make some adjustments here.

upload: ignore for now
like: favourite (schema doesn't yet exist)
comment: ignore for now
subscription: following
recommendation: favourite (schema doesn't yet exist)
playlistItem: ignore for now

I'll create the favourite schema and share with you today.

File: src/providers/google/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/index.ts:43]: `youtube.readonly`: is sufficient for read only purpose
`youtube`: is optional. might need for any activities require write permission through verida
`youtube.force-ssl`: to ensure a secure connection and confidentialtiy

Those are commonly used scopes.

We can add `youtube.readonly` only for now

File: src/providers/google/youtube-post.ts (hunk 3 above)
Comment (chime3) [reply src/providers/google/youtube-post.ts:193]: > Thanks, we will need to make some adjustments here.
> 
> upload: ignore for now like: favourite (schema doesn't yet exist) comment: ignore for now subscription: following recommendation: favourite (schema doesn't yet exist) playlistItem: ignore for now
> 
> I'll create the favourite schema and share with you today.

Okay

File: src/providers/google/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/index.ts:43]: > `youtube.readonly`: is sufficient for read only purpose `youtube`: is optional. might need for any activities require write permission through verida `youtube.force-ssl`: to ensure a secure connection and confidentialtiy
> 
> Those are commonly used scopes.
> 
> We can add `youtube.readonly` only for now

Please @tahpot

File: src/providers/google/youtube-post.ts (hunk 3 above)
Comment (tahpot) [reply src/providers/google/youtube-post.ts:193]: @chime3 See https://common.schemas.verida.io/favourite/v0.1.0/schema.json

File: src/providers/google/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tahpot) [reply src/providers/google/index.ts:43]: @chime3 Just use `youtube.readonly`

[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/index.ts:43]: > @chime3 Just use `youtube.readonly`

Yeah, will do

File: src/providers/google/youtube-post.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                    break;
+            }
+
+
+            results.push({
+                _id: `youtube-${postId}`,
+                name: title,
+                icon: iconUri,
+                uri: activityUri,
+                type: activityType,
```
Comment (tahpot) [inline src/providers/google/youtube-post.ts:173]: Type must match the `type` value from schema (https://common.schemas.verida.io/social/post/v0.1.0/schema.json).

ie: `video`

**Review by tahpot — CHANGES_REQUESTED** [review]

File: src/providers/google/youtube-following.ts [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+            const snippet = item.snippet;
+            const insertedAt = snippet.publishedAt || "Unknown";
+            
+            if (breakTimestamp && insertedAt < breakTimestamp) {
+                break;
+            }
+            
+            const title = snippet.title || "No title";
+            const description = snippet.description || "No description";
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/youtube-following.ts:153]: > Description isn't included in the item. Also check other handlers.

There is no matching field in `following` schema, but it is included in other handlers

File: src/providers/google/youtube-following.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    public getName(): string {
+        return "youtube-following";
+    }
+
+    public getSchemaUri(): string {
+        return CONFIG.verida.schemas.FOLLOWING;
+    }
+
+    public getProviderApplicationUrl(): string {
+        return "https://youtube.com";
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/youtube-following.ts:27]: > Add trailing `/` to match gmail.
> 
> `https://youtube.com/`. Check the others.

Done

File: src/providers/google/youtube-post.ts (hunk 5 above)
Comment (chime3) [reply src/providers/google/youtube-post.ts]: > Type must match the `type` value from schema (https://common.schemas.verida.io/social/post/v0.1.0/schema.json).
> 
> ie: `video`

Defined `postType`

File: src/providers/google/youtube-following.ts [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            const description = snippet.description || "No description"; 
+            const uri = "https://www.youtube.com/channel/" + snippet.resourceId.channelId;
+            const icon = snippet.thumbnails.default.url;
+
+            results.push({
+                _id: `youtube-${itemId}`,
+                name: title,
+                icon: icon,
+                uri: uri,
+                sourceData: snippet,
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/youtube-following.ts:163]: > `sourceId` isn't set, so tests fail.
> 
> Check others.

Done

File: src/providers/google/youtube-favourite.ts [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+                case SchemaYoutubeActivityType.RECOMMENDATION:
+                    videoId = contentDetails.recommendation.resourceId.videoId;
+                    activityUri = 'https://www.youtube.com/watch?v=' + videoId;
+                    break;
+                default:
+                    activityUri = 'Unknown activity type';
+                    break;
+            }
+
+            results.push({
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/youtube-favourite.ts:147]: > Have you tested this? I have liked a few videos and the tests show no data is being returned.

Syncing works now

**Review by tahpot — CHANGES_REQUESTED** [review]

File: src/providers/google/youtube-following.ts [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        breakId: string,
+        breakTimestamp?: string
+    ): Promise<SchemaFollowing[]> {
+        const results: SchemaFollowing[] = [];
+        for (const item of serverResponse.data.items) {
+            const itemId = `${this.connection.profile.id}-${item.id}`;
+            console.log(item)
+
+            if (itemId == breakId) {
+                const logEvent: SyncProviderLogEvent = {
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/youtube-following.ts:139]: I've added all

File: src/providers/google/youtube-following.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/youtube-following.ts:170]: > `sourceId` should only contain the ID from the source. It currently includes the `connection.profile.id` prefix.

Fixed

File: src/providers/google/youtube-following.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (tahpot) [reply src/providers/google/youtube-following.ts]: Ah, I've made a mistake.

All schemas inherit from the base schema (https://core.schemas.verida.io/base/v0.1.0/schema.json) which has a `summary` field.

For `following` and `favourite`, map `description` to the `summary` field.

File: src/providers/google/youtube-post.ts [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+        breakTimestamp?: string
+    ): Promise<SchemaPost[]> {
+        const results: SchemaPost[] = [];
+
+        const activities = serverResponse.data.items;
+        // filter post(upload, comment, bulletin)
+        const posts = activities.filter(activity => [SchemaYoutubeActivityType.UPLOAD, SchemaYoutubeActivityType.COMMENT].includes(activity.snippet.type as SchemaYoutubeActivityType))
+        for (const post of posts) {
+            const postId = post.id;
+            console.log(post)
```
Comment (tahpot) [inline src/providers/google/youtube-post.ts:139]: Remove console.log entries

File: src/providers/google/youtube-post.ts [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+
+            const title = snippet.title || "No title";
+            const description = snippet.description || "No description";
+            const contentDetails = post.contentDetails;
+
+            const activityType = snippet.type;
+            const iconUri = snippet.thumbnails.default.url;
+            // extract activity URI
+            let activityUri = "";
+            switch (activityType) {
```
Comment (tahpot) [inline src/providers/google/youtube-post.ts:170]: Only includ "Upload" type as a post. Ignore comments and other activities.

File: src/schemas.ts [hunk 13]
```diff
@@ -18,8 +18,22 @@ export interface SchemaFollowing extends SchemaRecord {
     insertedAt: string
 }
 
+export enum PostType {
```
Comment (tahpot) [inline src/schemas.ts:21]: SchemaPostType

File: src/schemas.ts [hunk 14]
```diff
@@ -49,4 +63,34 @@ export interface SchemaEmail extends SchemaRecord {
     sentAt: string
     attachments?: SchemaEmailAttachment[]
     threadId?: string
+}
+
+export enum SchemaYoutubeActivityType {
```
Comment (tahpot) [inline src/schemas.ts:68]: This isn't part of our schemas, so shouldn't be in this file.

Rename to `YoutubeActivityType` and put in a `providers/google/interfaces.ts` file as it's specific to the Google API.

File: src/schemas.ts [hunk 15]
```diff
@@ -49,4 +63,34 @@ export interface SchemaEmail extends SchemaRecord {
...
+    UPLOAD = "upload", // post
+    LIKE = "like", // favourite
+    SUBSCRIPTION = "subscription", // following
+    FAVOURITE = "favourite", // favourite
+    COMMENT = "comment", // post
+    PLAYLIST_ITEM = "playlistItem", //ignored
+    RECOMMENDATION = "recommendation", // favourite
+}
+
+export enum FavouriteType {
```
Comment (tahpot) [inline src/schemas.ts:78]: SchemaFavouriteType

File: src/schemas.ts [hunk 16]
```diff
@@ -49,4 +63,34 @@ export interface SchemaEmail extends SchemaRecord {
...
+    LIKE = "like",
+    FAVOURITE = "favourite",
+    RECOMMENDATION = "recommendation",
+    SHARE = "share"
+}
+
+export enum ContentType {
```
Comment (tahpot) [inline src/schemas.ts:85]: SchemaFavouriteContentType

File: src/providers/google/youtube-following.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/youtube-following.ts]: > Ah, I've made a mistake.
> 
> All schemas inherit from the base schema (https://core.schemas.verida.io/base/v0.1.0/schema.json) which has a `summary` field.
> 
> For `following` and `favourite`, map `description` to the `summary` field.

I will check

**Comment by tahpot** [comment]
@chime3 See my review comments.

The tests aren't passing. Please share screenshots of all tests passing when submitting this PR for review.

**Review by tahpot — CHANGES_REQUESTED** [review]
@chime3 See my review comments.

The tests aren't passing. Please share screenshots of all tests passing when submitting this PR for review.

File: src/providers/google/youtube-following.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/youtube-following.ts]: I modified `schemas.ts` file

File: src/schemas.ts (hunk 16 above)
Comment (chime3) [reply src/schemas.ts]: > SchemaFavouriteContentType

Fixed

File: src/schemas.ts (hunk 15 above)
Comment (chime3) [reply src/schemas.ts]: > SchemaFavouriteType

Fixed

File: src/schemas.ts (hunk 14 above)
Comment (chime3) [reply src/schemas.ts]: > This isn't part of our schemas, so shouldn't be in this file.
> 
> Rename to `YoutubeActivityType` and put in a `providers/google/interfaces.ts` file as it's specific to the Google API.

Fixed

File: src/schemas.ts (hunk 13 above)
Comment (chime3) [reply src/schemas.ts]: > SchemaPostType

Fixed

**Comment by chime3** [comment]
> @chime3 See my review comments.
> 
> The tests aren't passing. Please share screenshots of all tests passing when submitting this PR for review.

youtube-favourite-test
youtube-following-test
youtube-post-test
Modified `common.tests.ts` @tahpot

File: src/providers/google/youtube-favourite.ts [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+            // No results found, so stop sync
+            syncPosition = this.stopSync(syncPosition);
+    
+            return {
+                position: syncPosition,
+                results: [],
+            };
+        }
+        
+        // Sort items by publishedAt timestamp in descending order (most recent first)
```
Comment (tahpot) [inline src/providers/google/youtube-favourite.ts:66]: What's the purpose of this code?

It's only sorting locally.

Same question applies to the other youtube handlers.

**Review by tahpot — CHANGES_REQUESTED** [review]
## Test results

`youtube-post`: Appears to be working. I only have two youtube uploads so had to tweak the tests a bit locally

`youtube-following`: Failing: `AssertionError [ERR_ASSERTION]: Sync is still active`

`youtube-favourite`: Passing

## Sync

I tried syncing via the command line and the code produces save errors indicating the schema's aren't being populated correctly.

I have made a fix in the `develop` branch. Can you merge `develop`, resole any issues and then run sync from the command line to confirm data is being saved correctly.

## Other

I made some minor commits to improve the tests, see my commits.

**Comment by chime3** [comment]
> ## Test results
> `youtube-post`: Appears to be working. I only have two youtube uploads so had to tweak the tests a bit locally
> 
> `youtube-following`: Failing: `AssertionError [ERR_ASSERTION]: Sync is still active`
> 
> `youtube-favourite`: Passing
> 
> ## Sync
> I tried syncing via the command line and the code produces save errors indicating the schema's aren't being populated correctly.
> 
> I have made a fix in the `develop` branch. Can you merge `develop`, resole any issues and then run sync from the command line to confirm data is being saved correctly.
> 
> ## Other
> I made some minor commits to improve the tests, see my commits.

I got it, thank you

File: src/providers/google/youtube-favourite.ts (hunk 17 above)
Comment (chime3) [reply src/providers/google/youtube-favourite.ts:66]: Yeah, it is for local sort.
Do you want to remove them? @tahpot

**Comment by chime3** [comment]
> ## Test results
> `youtube-post`: Appears to be working. I only have two youtube uploads so had to tweak the tests a bit locally
> 
> `youtube-following`: Failing: `AssertionError [ERR_ASSERTION]: Sync is still active`
> 
> `youtube-favourite`: Passing
> 
> ## Sync
> I tried syncing via the command line and the code produces save errors indicating the schema's aren't being populated correctly.
> 
> I have made a fix in the `develop` branch. Can you merge `develop`, resole any issues and then run sync from the command line to confirm data is being saved correctly.
> 
> ## Other
> I made some minor commits to improve the tests, see my commits.

I merged `develop` into `feature/youtube-handler` and run test passing.
image

Please try to reset connection if `AssertionError [ERR_ASSERTION]: Sync is still active` error persists.

File: src/providers/google/youtube-favourite.ts (hunk 17 above)
Comment (chime3) [reply src/providers/google/youtube-favourite.ts:66]: I removed the sort.

**Comment by tahpot** [comment]
Thanks, here's my feedback:

1. `youtube-following` now works. The issue was I only had 6 subscriptions, but needed 7 for a third page of results to exist and the sync to remain set to "active". Can you add notes on what you need to do to your youtube account for the tests to pass in the `providers/google/README.md` file? ie: Youtube following needs at least 7 subscriptions, with one created within the last 24 hrs for tests to pass.
2. Sync is now saving data
3. There is a bug with all the handlers not tracking the position correctly. If you run the sync multiple times, it re-processes the existing records and re-saves them. The purpose of the sync position is to only process unprocessed items.

**Comment by chime3** [comment]
> Thanks, here's my feedback:
> 
> 1. `youtube-following` now works. The issue was I only had 6 subscriptions, but needed 7 for a third page of results to exist and the sync to remain set to "active". Can you add notes on what you need to do to your youtube account for the tests to pass in the `providers/google/README.md` file? ie: Youtube following needs at least 7 subscriptions, with one created within the last 24 hrs for tests to pass.

I added `Prerequisites` in `README.md` file.

**Comment by chime3** [comment]
> 2. Sync is now saving data
> 3. There is a bug with all the handlers not tracking the position correctly. If you run the sync multiple times, it re-processes the existing records and re-saves them. The purpose of the sync position is to only process unprocessed items.

I referred to `gmail` handler. Is `gmail` sync not implemented correctly as well? @tahpot

**Comment by tahpot** [comment]
> > 2. Sync is now saving data
> > 3. There is a bug with all the handlers not tracking the position correctly. If you run the sync multiple times, it re-processes the existing records and re-saves them. The purpose of the sync position is to only process unprocessed items.
> 
> I referred to `gmail` handler. Is `gmail` sync not implemented correctly as well? @tahpot

Good question!

I haven't actually reached the end of my gmail, so it's possible it doesn't work correctly and I hadn't yet noticed.

I'll approve this PR and we can address this problem in the Google provider separately.

I have built a different process for telegram which I will port across to gmail, youtube and then make sure this problem is also solved.

**Review by tahpot — APPROVED** [review]
There is a conflict that needs to be fixed, then this is approved for merging

**Comment by tahpot** [comment]
@chime3 I have made some changes to the underlying schemas.

Please review these changes I have made in another branch and update the youtube handlers to match:

1. Add a `syncMessage` into the handler that will be displayed to users to help them understand the current state of processing. https://github.com/verida/data-connector-server/pull/90/commits/3bf6aa5aba27b68afc2b42415e6295962b07f675
2. The enum status for the progress have changed, see here: https://github.com/verida/data-connector-server/pull/90/commits/ccf8c973d632f0c31315aac2f01f0983263e8814

Note: You may want to merge my branch into your branch so you get the new enum values.

**Comment by tahpot** [comment]
> > 2. Sync is now saving data
> > 3. There is a bug with all the handlers not tracking the position correctly. If you run the sync multiple times, it re-processes the existing records and re-saves them. The purpose of the sync position is to only process unprocessed items.
> 
> I referred to `gmail` handler. Is `gmail` sync not implemented correctly as well? @tahpot

It turns out there was a bug in the core handler that caused this issue which I have now fixed in the other branch.

If you merge the branch (as mentioned above) it should fix the issue.... however you will also need to delete all your connections and start again as they will be corrupt with incorrect data caused by the bug.

**Comment by chime3** [comment]
> There is a conflict that needs to be fixed, then this is approved for merging

Solved

**Comment by tahpot** [comment]
> > There is a conflict that needs to be fixed, then this is approved for merging
> 
> Solved

You need to merge this branch: https://github.com/verida/data-connector-server/issues/89

**Comment by chime3** [comment]
> > > There is a conflict that needs to be fixed, then this is approved for merging
> > 
> > 
> > Solved
> 
> You need to merge this branch: #89

Yeah, I got it

**Comment by chime3** [comment]
> > > > 2. Sync is now saving data
> > > > 3. There is a bug with all the handlers not tracking the position correctly. If you run the sync multiple times, it re-processes the existing records and re-saves them. The purpose of the sync position is to only process unprocessed items.
> > > 
> > > 
> > > I referred to `gmail` handler. Is `gmail` sync not implemented correctly as well? @tahpot
> > 
> > 
> > It turns out there was a bug in the core handler that caused this issue which I have now fixed in the other branch.
> > If you merge the branch (as mentioned above) it should fix the issue.... however you will also need to delete all your connections and start again as they will be corrupt with incorrect data caused by the bug.
> 
> I can confirm that my fix now means the youtube handlers don't save the same item more than once.

Okay, let me check after merge

**Comment by chime3** [comment]
> > > > 2. Sync is now saving data
> > > > 3. There is a bug with all the handlers not tracking the position correctly. If you run the sync multiple times, it re-processes the existing records and re-saves them. The purpose of the sync position is to only process unprocessed items.
> > > 
> > > 
> > > I referred to `gmail` handler. Is `gmail` sync not implemented correctly as well? @tahpot
> > 
> > 
> > It turns out there was a bug in the core handler that caused this issue which I have now fixed in the other branch.
> > If you merge the branch (as mentioned above) it should fix the issue.... however you will also need to delete all your connections and start again as they will be corrupt with incorrect data caused by the bug.
> 
> I can confirm that my fix now means the youtube handlers don't save the same item more than once.

Handlers save items only once now.

**Review by tahpot — APPROVED** [review]
Approved!

LFG

**Pull request merged**